pkg/controller/installation/core_controller.go

```diff
@@ -411,38 +416,32 @@ func updateInstallationWithDefaults(ctx context.Context, client client.Client, i
 		awsNode = nil
 	}
 
-	err = mergeAndFillDefaults(instance, openshiftConfig, kubeadmConfig, awsNode)
+	currentPools, err := getActivePools(ctx, client)
+	if err != nil {
+		return fmt.Errorf("unable to list IPPools: %s", err.Error())
+	}
+
+	err = MergeAndFillDefaults(instance, awsNode, currentPools)
 	if err != nil {
 		return err
 	}
 	return nil
 }
 
-// mergeAndFillDefaults merges in configuration from the Kubernetes provider, if applicable, and then
+// MergeAndFillDefaults merges in configuration from the Kubernetes provider, if applicable, and then
 // populates defaults in the Installation instance.
-func mergeAndFillDefaults(i *operator.Installation, o *configv1.Network, kubeadmConfig *corev1.ConfigMap, awsNode *appsv1.DaemonSet) error {
-	if o != nil {
-		// Merge in OpenShift configuration.
-		if err := updateInstallationForOpenshiftNetwork(i, o); err != nil {
-			return fmt.Errorf("Could not resolve CalicoNetwork IPPool and OpenShift network: %s", err.Error())
-		}
-	} else if kubeadmConfig != nil {
-		// Merge in kubeadm configuration.
-		if err := updateInstallationForKubeadm(i, kubeadmConfig); err != nil {
-			return fmt.Errorf("Could not resolve CalicoNetwork IPPool and kubeadm configuration: %s", err.Error())
-		}
-	}
+func MergeAndFillDefaults(i *operator.Installation, awsNode *appsv1.DaemonSet, currentPools *crdv1.IPPoolList) error {
 	if awsNode != nil {
 		if err := updateInstallationForAWSNode(i, awsNode); err != nil {
 			return fmt.Errorf("Could not resolve AWS node configuration: %s", err.Error())
 		}
 	}
 
-	return fillDefaults(i)
+	return fillDefaults(i, currentPools)
 }
 
 // fillDefaults populates the default values onto an Installation object.
-func fillDefaults(instance *operator.Installation) error {
+func fillDefaults(instance *operator.Installation, currentPools *crdv1.IPPoolList) error {
```

Comment:
We need to pass the current set of IP pools, as read from the cluster's API server, so that Installation defaulting can fill in things like IP auto detection which we do based on configured pools.

This is a notable change in behavior - previously, we only did this based on the pools in the installation. But now we do it based on the IP pools in the cluster as a whole. i.e., if a pool was created by calicoctl, etc. 

I think this is OK enough, given we have written previous defaults back to the Installation on any existing clusters so it won't break upgrade, and any IP pools created after Calico is started will be ignored by this defaulting anyway for the same reason.

Reply:
We could of course switch this to continue using the IP pools in the Installation though, in order to maintain the behavior that the operator simply ignores IP pools that it itself didn't create.

Reply:
> It's also probably on average the more desirable behavior.

Could you elaborate on why this is more desirable?

Reply:
In most cases, I think this is functionally equivalent to how it behaved before - the IP pools in the cluster should match those in the Installation exactly.

However, the one scenario where this is more desirable is the case where users want to manage IP pools by hand - e.g., using calicoctl - and disable IP pool management in the operator entirely. Without this, that scenario doesn't default the Installation resource as correctly. It will see no IP pools in the InstallationSpec and assume that IP autodetection should be disabled and IPAM assignment should be disabled.

It's niche, but I think the right behavior in that niche case.

For all other cases, I believe this is largely the same. If at least one IP pool of all address families in-use on the cluster is defined in the Installation, then the defaulting result will be the same regardless of this change.

Reply:
Makes sense, sounds like an improvement overall.